File: week-2/scripts/common/schema/types/amm.ts
/**
 * Program IDL in camelCase format in order to be used in JS/TS.
 *
 * Note that this is only a type helper and is not the actual IDL. The original
 * IDL can be found at `target/idl/amm.json`.
 */
export type Amm = {
  "address": "CpuYGzAZWKWBHXUoBSfEg3qnvRd8pMcRa9XV29Xoj3KU",
  "metadata": {
    "name": "amm",
    "version": "1.0.0",
    "spec": "0.1.0",
    "description": "Created with Anchor",
    "repository": "https://github.com/M-Daeva/solana-boilerplate"
  },
  "instructions": [
    {
      "name": "createPool",
      "discriminator": [
        233,
        146,
        209,
        142,
        207,
        104,
        64,
        188
      ],
      "accounts": [
        {
          "name": "systemProgram",
          "address": "11111111111111111111111111111111"
        },
        {
          "name": "tokenProgram"
        },
        {
          "name": "associatedTokenProgram",
          "address": "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL"
        },
        {
          "name": "poolCreator",
          "writable": true,
          "signer": true
        },
        {
          "name": "poolConfig",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "const",
                "value": [
                  99,
                  111,
                  110,
                  102,
                  105,
                  103
                ]
              },
              {
                "kind": "arg",
                "path": "id"
              }
            ]
          }
        },
        {
          "name": "poolBalance",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "const",
                "value": [
                  98,
                  97,
                  108,
                  97,
                  110,
                  99,
                  101
                ]
              },
              {
                "kind": "arg",
                "path": "id"
              }
            ]
          }
        },
        {
          "name": "mintLp",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "const",
                "value": [
                  108,
                  112
                ]
              },
              {
                "kind": "arg",
                "path": "id"
              }
            ]
          }
        },
        {
          "name": "mintX"
        },
        {
          "name": "mintY"
        },
        {
          "name": "liquidityPoolMintLpAta",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "path": "poolConfig"
              },
              {
                "kind": "account",
                "path": "tokenProgram"
              },
              {
                "kind": "account",
                "path": "mintLp"
              }
            ],
            "program": {
              "kind": "const",
              "value": [
                140,
                151,
                37,
                143,
                78,
                36,
                137,
                241,
                187,
                61,
                16,
                41,
                20,
                142,
                13,
                131,
                11,
                90,
                19,
                153,
                218,
                255,
                16,
                132,
                4,
                142,
                123,
                216,
                219,
                233,
                248,
                89
              ]
            }
          }
        },
        {
          "name": "liquidityPoolMintXAta",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "path": "poolConfig"
              },
              {
                "kind": "account",
                "path": "tokenProgram"
              },
              {
                "kind": "account",
                "path": "mintX"
              }
            ],
            "program": {
              "kind": "const",
              "value": [
                140,
                151,
                37,
                143,
                78,
                36,
                137,
                241,
                187,
                61,
                16,
                41,
                20,
                142,
                13,
                131,
                11,
                90,
                19,
                153,
                218,
                255,
                16,
                132,
                4,
                142,
                123,
                216,
                219,
                233,
                248,
                89
              ]
            }
          }
        },
        {
          "name": "liquidityPoolMintYAta",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "path": "poolConfig"
              },
              {
                "kind": "account",
                "path": "tokenProgram"
              },
              {
                "kind": "account",
                "path": "mintY"
              }
            ],
            "program": {
              "kind": "const",
              "value": [
                140,
                151,
                37,
                143,
                78,
                36,
                137,
                241,
                187,
                61,
                16,
                41,
                20,
                142,
                13,
                131,
                11,
                90,
                19,
                153,
                218,
                255,
                16,
                132,
                4,
                142,
                123,
                216,
                219,
                233,
                248,
                89
              ]
            }
          }
        }
      ],
      "args": [
        {
          "name": "id",
          "type": "u64"
        },
        {
          "name": "mintX",
          "type": "pubkey"
        },
        {
          "name": "mintY",
          "type": "pubkey"
        },
        {
          "name": "feeBps",
          "type": "u16"
        }
      ]
    },
    {
      "name": "provideLiquidity",
      "discriminator": [
        40,
        110,
        107,
        116,
        174,
        127,
        97,
        204
      ],
      "accounts": [
        {
          "name": "systemProgram",
          "address": "11111111111111111111111111111111"
        },
        {
          "name": "tokenProgram"
        },
        {
          "name": "associatedTokenProgram",
          "address": "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL"
        },
        {
          "name": "liquidityProvider",
          "writable": true,
          "signer": true
        },
        {
          "name": "poolConfig",
          "pda": {
            "seeds": [
              {
                "kind": "const",
                "value": [
                  99,
                  111,
                  110,
                  102,
                  105,
                  103
                ]
              },
              {
                "kind": "arg",
                "path": "id"
              }
            ]
          }
        },
        {
          "name": "poolBalance",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "const",
                "value": [
                  98,
                  97,
                  108,
                  97,
                  110,
                  99,
                  101
                ]
              },
              {
                "kind": "arg",
                "path": "id"
              }
            ]
          }
        },
        {
          "name": "mintLp",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "const",
                "value": [
                  108,
                  112
                ]
              },
              {
                "kind": "arg",
                "path": "id"
              }
            ]
          }
        },
        {
          "name": "mintX"
        },
        {
          "name": "mintY"
        },
        {
          "name": "liquidityProviderMintLpAta",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "path": "liquidityProvider"
              },
              {
                "kind": "account",
                "path": "tokenProgram"
              },
              {
                "kind": "account",
                "path": "mintLp"
              }
            ],
            "program": {
              "kind": "const",
              "value": [
                140,
                151,
                37,
                143,
                78,
                36,
                137,
                241,
                187,
                61,
                16,
                41,
                20,
                142,
                13,
                131,
                11,
                90,
                19,
                153,
                218,
                255,
                16,
                132,
                4,
                142,
                123,
                216,
                219,
                233,
                248,
                89
              ]
            }
          }
        },
        {
          "name": "liquidityProviderMintXAta",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "path": "liquidityProvider"
              },
              {
                "kind": "account",
                "path": "tokenProgram"
              },
              {
                "kind": "account",
                "path": "mintX"
              }
            ],
            "program": {
              "kind": "const",
              "value": [
                140,
                151,
                37,
                143,
                78,
                36,
                137,
                241,
                187,
                61,
                16,
                41,
                20,
                142,
                13,
                131,
                11,
                90,
                19,
                153,
                218,
                255,
                16,
                132,
                4,
                142,
                123,
                216,
                219,
                233,
                248,
                89
              ]
            }
          }
        },
        {
          "name": "liquidityProviderMintYAta",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "path": "liquidityProvider"
              },
              {
                "kind": "account",
                "path": "tokenProgram"
              },
              {
                "kind": "account",
                "path": "mintY"
              }
            ],
            "program": {
              "kind": "const",
              "value": [
                140,
                151,
                37,
                143,
                78,
                36,
                137,
                241,
                187,
                61,
                16,
                41,
                20,
                142,
                13,
                131,
                11,
                90,
                19,
                153,
                218,
                255,
                16,
                132,
                4,
                142,
                123,
                216,
                219,
                233,
                248,
                89
              ]
            }
          }
        },
        {
          "name": "liquidityPoolMintLpAta",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "path": "poolConfig"
              },
              {
                "kind": "account",
                "path": "tokenProgram"
              },
              {
                "kind": "account",
                "path": "mintLp"
              }
            ],
            "program": {
              "kind": "const",
              "value": [
                140,
                151,
                37,
                143,
                78,
                36,
                137,
                241,
                187,
                61,
                16,
                41,
                20,
                142,
                13,
                131,
                11,
                90,
                19,
                153,
                218,
                255,
                16,
                132,
                4,
                142,
                123,
                216,
                219,
                233,
                248,
                89
              ]
            }
          }
        },
        {
          "name": "liquidityPoolMintXAta",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "path": "poolConfig"
              },
              {
                "kind": "account",
                "path": "tokenProgram"
              },
              {
                "kind": "account",
                "path": "mintX"
              }
            ],
            "program": {
              "kind": "const",
              "value": [
                140,
                151,
                37,
                143,
                78,
                36,
                137,
                241,
                187,
                61,
                16,
                41,
                20,
                142,
                13,
                131,
                11,
                90,
                19,
                153,
                218,
                255,
                16,
                132,
                4,
                142,
                123,
                216,
                219,
                233,
                248,
                89
              ]
            }
          }
        },
        {
          "name": "liquidityPoolMintYAta",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "path": "poolConfig"
              },
              {
                "kind": "account",
                "path": "tokenProgram"
              },
              {
                "kind": "account",
                "path": "mintY"
              }
            ],
            "program": {
              "kind": "const",
              "value": [
                140,
                151,
                37,
                143,
                78,
                36,
                137,
                241,
                187,
                61,
                16,
                41,
                20,
                142,
                13,
                131,
                11,
                90,
                19,
                153,
                218,
                255,
                16,
                132,
                4,
                142,
                123,
                216,
                219,
                233,
                248,
                89
              ]
            }
          }
        }
      ],
      "args": [
        {
          "name": "id",
          "type": "u64"
        },
        {
          "name": "mintXAmount",
          "type": "u64"
        },
        {
          "name": "mintYAmount",
          "type": "u64"
        }
      ]
    },
    {
      "name": "swap",
      "discriminator": [
        248,
        198,
        158,
        145,
        225,
        117,
        135,
        200
      ],
      "accounts": [
        {
          "name": "systemProgram",
          "address": "11111111111111111111111111111111"
        },
        {
          "name": "tokenProgram"
        },
        {
          "name": "associatedTokenProgram",
          "address": "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL"
        },
        {
          "name": "trader",
          "writable": true,
          "signer": true
        },
        {
          "name": "poolConfig",
          "pda": {
            "seeds": [
              {
                "kind": "const",
                "value": [
                  99,
                  111,
                  110,
                  102,
                  105,
                  103
                ]
              },
              {
                "kind": "arg",
                "path": "id"
              }
            ]
          }
        },
        {
          "name": "poolBalance",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "const",
                "value": [
                  98,
                  97,
                  108,
                  97,
                  110,
                  99,
                  101
                ]
              },
              {
                "kind": "arg",
                "path": "id"
              }
            ]
          }
        },
        {
          "name": "mintX"
        },
        {
          "name": "mintY"
        },
        {
          "name": "traderMintXAta",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "path": "trader"
              },
              {
                "kind": "account",
                "path": "tokenProgram"
              },
              {
                "kind": "account",
                "path": "mintX"
              }
            ],
            "program": {
              "kind": "const",
              "value": [
                140,
                151,
                37,
                143,
                78,
                36,
                137,
                241,
                187,
                61,
                16,
                41,
                20,
                142,
                13,
                131,
                11,
                90,
                19,
                153,
                218,
                255,
                16,
                132,
                4,
                142,
                123,
                216,
                219,
                233,
                248,
                89
              ]
            }
          }
        },
        {
          "name": "traderMintYAta",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "path": "trader"
              },
              {
                "kind": "account",
                "path": "tokenProgram"
              },
              {
                "kind": "account",
                "path": "mintY"
              }
            ],
            "program": {
              "kind": "const",
              "value": [
                140,
                151,
                37,
                143,
                78,
                36,
                137,
                241,
                187,
                61,
                16,
                41,
                20,
                142,
                13,
                131,
                11,
                90,
                19,
                153,
                218,
                255,
                16,
                132,
                4,
                142,
                123,
                216,
                219,
                233,
                248,
                89
              ]
            }
          }
        },
        {
          "name": "liquidityPoolMintXAta",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "path": "poolConfig"
              },
              {
                "kind": "account",
                "path": "tokenProgram"
              },
              {
                "kind": "account",
                "path": "mintX"
              }
            ],
            "program": {
              "kind": "const",
              "value": [
                140,
                151,
                37,
                143,
                78,
                36,
                137,
                241,
                187,
                61,
                16,
                41,
                20,
                142,
                13,
                131,
                11,
                90,
                19,
                153,
                218,
                255,
                16,
                132,
                4,
                142,
                123,
                216,
                219,
                233,
                248,
                89
              ]
            }
          }
        },
        {
          "name": "liquidityPoolMintYAta",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "account",
                "path": "poolConfig"
              },
              {
                "kind": "account",
                "path": "tokenProgram"
              },
              {
                "kind": "account",
                "path": "mintY"
              }
            ],
            "program": {
              "kind": "const",
              "value": [
                140,
                151,
                37,
                143,
                78,
                36,
                137,
                241,
                187,
                61,
                16,
                41,
                20,
                142,
                13,
                131,
                11,
                90,
                19,
                153,
                218,
                255,
                16,
                132,
                4,
                142,
                123,
                216,
                219,
                233,
                248,
                89
              ]
            }
          }
        }
      ],
      "args": [
        {
          "name": "id",
          "type": "u64"
        },
        {
          "name": "amountIn",
          "type": "u64"
        },
        {
          "name": "mintIn",
          "type": "pubkey"
        }
      ]
    }
  ],
  "accounts": [
    {
      "name": "poolBalance",
      "discriminator": [
        34,
        231,
        160,
        225,
        133,
        53,
        84,
        229
      ]
    },
    {
      "name": "poolConfig",
      "discriminator": [
        26,
        108,
        14,
        123,
        116,
        230,
        129,
        43
      ]
    }
  ],
  "errors": [
    {
      "code": 6000,
      "name": "noLiquidity",
      "msg": "No liquidity is provided"
    },
    {
      "code": 6001,
      "name": "wrongMint",
      "msg": "The mint isn't supported"
    }
  ],
  "types": [
    {
      "name": "poolBalance",
      "type": {
        "kind": "struct",
        "fields": [
          {
            "name": "mintXAmount",
            "type": "u64"
          },
          {
            "name": "mintYAmount",
            "type": "u64"
          },
          {
            "name": "mintLpAmount",
            "type": "u64"
          }
        ]
      }
    },
    {
      "name": "poolConfig",
      "type": {
        "kind": "struct",
        "fields": [
          {
            "name": "configBump",
            "type": "u8"
          },
          {
            "name": "balanceBump",
            "type": "u8"
          },
          {
            "name": "lpBump",
            "type": "u8"
          },
          {
            "name": "id",
            "type": "u64"
          },
          {
            "name": "authority",
            "type": {
              "option": "pubkey"
            }
          },
          {
            "name": "mintX",
            "type": "pubkey"
          },
          {
            "name": "mintY",
            "type": "pubkey"
          },
          {
            "name": "mintLp",
            "type": "pubkey"
          },
          {
            "name": "feeBps",
            "type": "u16"
          },
          {
            "name": "isLocked",
            "type": "bool"
          }
        ]
      }
    }
  ]
};

File: week-2/tests-rs/src/helpers/amm.rs
use {
    crate::helpers::suite::{
        core::{
            extension::{get_data, send_tx_with_ix},
            token::WithTokenKeys,
            App, ProgramId,
        },
        types::{AppToken, AppUser},
    },
    amm::{accounts, instruction, state},
    anchor_lang::Result,
    litesvm::types::TransactionMetadata,
};

pub trait AmmExtension {
    fn amm_try_create_pool(
        &mut self,
        sender: AppUser,
        id: u64,
        mint_x: AppToken,
        mint_y: AppToken,
        fee_bps: u16,
    ) -> Result<TransactionMetadata>;

    fn amm_try_provide_liquidity(
        &mut self,
        sender: AppUser,
        id: u64,
        mint_x_amount: u64,
        mint_y_amount: u64,
    ) -> Result<TransactionMetadata>;

    fn amm_try_withdraw_liquidity(
        &mut self,
        sender: AppUser,
        id: u64,
        mint_lp_amount: u64,
    ) -> Result<TransactionMetadata>;

    fn amm_try_swap(
        &mut self,
        sender: AppUser,
        id: u64,
        amount_in: u64,
        mint_in: AppToken,
    ) -> Result<TransactionMetadata>;

    fn amm_query_pool_config(&self, pool_id: u64) -> Result<state::PoolConfig>;

    fn amm_query_pool_config_list(&self) -> Result<Vec<state::PoolConfig>>;

    fn amm_query_pool_balance(&self, pool_id: u64) -> Result<state::PoolBalance>;
}

impl AmmExtension for App {
    fn amm_try_create_pool(
        &mut self,
        sender: AppUser,
        id: u64,
        mint_x: AppToken,
        mint_y: AppToken,
        fee_bps: u16,
    ) -> Result<TransactionMetadata> {
        // programs
        let ProgramId {
            amm: program_id,
            system_program,
            token_program,
            associated_token_program,
        } = self.program_id;

        // signers
        let payer = sender.pubkey();
        let signers = [sender.keypair()];

        // mint
        let mint_x = mint_x.pubkey(&self);
        let mint_y = mint_y.pubkey(&self);

        // pda
        let pool_config = self.pda.amm_pool_config(id);
        let pool_balance = self.pda.amm_pool_balance(id);
        let mint_lp = self.pda.amm_mint_lp(id);

        // ata
        let liquidity_pool_mint_lp_ata = App::get_ata(&pool_config, &mint_lp);
        let liquidity_pool_mint_x_ata = App::get_ata(&pool_config, &mint_x);
        let liquidity_pool_mint_y_ata = App::get_ata(&pool_config, &mint_y);

        let accounts = accounts::CreatePool {
            system_program,
            token_program,
            associated_token_program,
            pool_creator: payer,
            pool_config,
            pool_balance,
            mint_lp,
            mint_x,
            mint_y,
            liquidity_pool_mint_lp_ata,
            liquidity_pool_mint_x_ata,
            liquidity_pool_mint_y_ata,
        };

        let instruction_data = instruction::CreatePool {
            id,
            mint_x,
            mint_y,
            fee_bps,
        };

        send_tx_with_ix(
            self,
            &program_id,
            &accounts,
            &instruction_data,
            &payer,
            &signers,
        )
    }

    fn amm_try_provide_liquidity(
        &mut self,
        sender: AppUser,
        id: u64,
        mint_x_amount: u64,
        mint_y_amount: u64,
    ) -> Result<TransactionMetadata> {
        // programs
        let ProgramId {
            amm: program_id,
            system_program,
            token_program,
            associated_token_program,
        } = self.program_id;

        // signers
        let payer = sender.pubkey();
        let signers = [sender.keypair()];

        // mint
        let state::PoolConfig { mint_x, mint_y, .. } = self.amm_query_pool_config(id)?;

        // pda
        let pool_config = self.pda.amm_pool_config(id);
        let pool_balance = self.pda.amm_pool_balance(id);
        let mint_lp = self.pda.amm_mint_lp(id);

        // ata
        let liquidity_pool_mint_lp_ata = App::get_ata(&pool_config, &mint_lp);
        let liquidity_pool_mint_x_ata = App::get_ata(&pool_config, &mint_x);
        let liquidity_pool_mint_y_ata = App::get_ata(&pool_config, &mint_y);

        let liquidity_provider_mint_lp_ata = App::get_ata(&payer, &mint_lp);
        let liquidity_provider_mint_x_ata = App::get_ata(&payer, &mint_x);
        let liquidity_provider_mint_y_ata = App::get_ata(&payer, &mint_y);

        let accounts = accounts::Liquidity {
            system_program,
            token_program,
            associated_token_program,
            liquidity_provider: payer,
            pool_config,
            pool_balance,
            mint_lp,
            mint_x,
            mint_y,
            liquidity_pool_mint_lp_ata,
            liquidity_pool_mint_x_ata,
            liquidity_pool_mint_y_ata,
            liquidity_provider_mint_lp_ata,
            liquidity_provider_mint_x_ata,
            liquidity_provider_mint_y_ata,
        };

        let instruction_data = instruction::ProvideLiquidity {
            _id: id,
            mint_x_amount,
            mint_y_amount,
        };

        send_tx_with_ix(
            self,
            &program_id,
            &accounts,
            &instruction_data,
            &payer,
            &signers,
        )
    }

    fn amm_try_withdraw_liquidity(
        &mut self,
        sender: AppUser,
        id: u64,
        mint_lp_amount: u64,
    ) -> Result<TransactionMetadata> {
        // programs
        let ProgramId {
            amm: program_id,
            system_program,
            token_program,
            associated_token_program,
        } = self.program_id;

        // signers
        let payer = sender.pubkey();
        let signers = [sender.keypair()];

        // mint
        let state::PoolConfig { mint_x, mint_y, .. } = self.amm_query_pool_config(id)?;

        // pda
        let pool_config = self.pda.amm_pool_config(id);
        let pool_balance = self.pda.amm_pool_balance(id);
        let mint_lp = self.pda.amm_mint_lp(id);

        // ata
        let liquidity_pool_mint_lp_ata = App::get_ata(&pool_config, &mint_lp);
        let liquidity_pool_mint_x_ata = App::get_ata(&pool_config, &mint_x);
        let liquidity_pool_mint_y_ata = App::get_ata(&pool_config, &mint_y);

        let liquidity_provider_mint_lp_ata = App::get_ata(&payer, &mint_lp);
        let liquidity_provider_mint_x_ata = App::get_ata(&payer, &mint_x);
        let liquidity_provider_mint_y_ata = App::get_ata(&payer, &mint_y);

        let accounts = accounts::Liquidity {
            system_program,
            token_program,
            associated_token_program,
            liquidity_provider: payer,
            pool_config,
            pool_balance,
            mint_lp,
            mint_x,
            mint_y,
            liquidity_pool_mint_lp_ata,
            liquidity_pool_mint_x_ata,
            liquidity_pool_mint_y_ata,
            liquidity_provider_mint_lp_ata,
            liquidity_provider_mint_x_ata,
            liquidity_provider_mint_y_ata,
        };

        let instruction_data = instruction::WithdrawLiquidity {
            _id: id,
            mint_lp_amount,
        };

        send_tx_with_ix(
            self,
            &program_id,
            &accounts,
            &instruction_data,
            &payer,
            &signers,
        )
    }

    fn amm_try_swap(
        &mut self,
        sender: AppUser,
        id: u64,
        amount_in: u64,
        mint_in: AppToken,
    ) -> Result<TransactionMetadata> {
        // programs
        let ProgramId {
            amm: program_id,
            system_program,
            token_program,
            associated_token_program,
        } = self.program_id;

        // signers
        let payer = sender.pubkey();
        let signers = [sender.keypair()];

        // mint
        let mint_in = mint_in.pubkey(&self);
        let state::PoolConfig { mint_x, mint_y, .. } = self.amm_query_pool_config(id)?;

        // pda
        let pool_config = self.pda.amm_pool_config(id);
        let pool_balance = self.pda.amm_pool_balance(id);

        // ata
        let liquidity_pool_mint_x_ata = App::get_ata(&pool_config, &mint_x);
        let liquidity_pool_mint_y_ata = App::get_ata(&pool_config, &mint_y);

        let trader_mint_x_ata = App::get_ata(&payer, &mint_x);
        let trader_mint_y_ata = App::get_ata(&payer, &mint_y);

        let accounts = accounts::Swap {
            system_program,
            token_program,
            associated_token_program,
            trader: payer,
            pool_config,
            pool_balance,
            mint_x,
            mint_y,
            liquidity_pool_mint_x_ata,
            liquidity_pool_mint_y_ata,
            trader_mint_x_ata,
            trader_mint_y_ata,
        };

        let instruction_data = instruction::Swap {
            _id: id,
            amount_in,
            mint_in,
        };

        send_tx_with_ix(
            self,
            &program_id,
            &accounts,
            &instruction_data,
            &payer,
            &signers,
        )
    }

    fn amm_query_pool_config(&self, pool_id: u64) -> Result<state::PoolConfig> {
        get_data(&self.litesvm, &self.pda.amm_pool_config(pool_id))
    }

    fn amm_query_pool_config_list(&self) -> Result<Vec<state::PoolConfig>> {
        unimplemented!()
    }

    fn amm_query_pool_balance(&self, pool_id: u64) -> Result<state::PoolBalance> {
        get_data(&self.litesvm, &self.pda.amm_pool_balance(pool_id))
    }
}
